### PR TITLE
Change unit annotations to `Quantity` annotations

### PR DESCRIPTION
**Pull request by namurphy**
## Description

This PR changes annotations like `u.kg` to `u.Quantity[u.kg]` so that they are true type hint annotations:

```python
import astropy.units as u

# old style unit annotations
@validate_quantities
def magnetic_pressure(B: u.T) -> u.Pa:
    ...

# new style with actual type hint annotations
@validate_quantities
def magnetic_pressure(B: u.Quantity(u.T)) -> u.Quantity(u.Pa):
    ...
```

## Motivation and context

At present, we use annotations like `u.kg` to specify the required units for `@validate_quantities`, which then verifies that the quantities are of the appropriate physical type and converts them to the correct units.  However, this prevents us from using static type checking tools like [mypy](https://mypy-lang.org). While these are annotations, they are not *type hint* annotations because things like `u.kg` are not a type.

Enabling mypy (#2424) will let us catch errors in type hint annotations, which we currently have no way of doing.  

Here's an example docstring [before (with unit annotations)](https://docs.plasmapy.org/en/v2023.10.x/api/plasmapy.formulary.collisions.dimensionless.coupling_parameter.html#plasmapy.formulary.collisions.dimensionless.coupling_parameter) and [after (with `Quantity` annotations)](https://plasmapy--2421.org.readthedocs.build/en/2421/api/plasmapy.formulary.collisions.dimensionless.coupling_parameter.html#coupling-parameter).

## Related issues

Related to: #268, #2424

### CODE REVIEW

File: docs/contributing/coding_guide.rst
Comment (namurphy) [inline docs/contributing/coding_guide.rst:1]: I switched the order of these paragraphs, but didn't make any substantive changes to the contents besides the annotations.

File: plasmapy/dispersion/analytical/mhd_waves_.py [hunk 1]
```diff
@@ -112,7 +112,9 @@ def beta(self):
 
     @staticmethod
     @validate_quantities
-    def _validate_k_theta(k: u.rad / u.m, theta: u.rad) -> u.Quantity:
+    def _validate_k_theta(
+        k: u.Quantity[u.rad / u.m], theta: u.Quantity[u.rad]
+    ) -> u.Quantity:
```
Comment (namurphy) [inline plasmapy/dispersion/analytical/mhd_waves_.py:117]: - [ ] Add a physical type here?

File: plasmapy/simulation/abstractions.py [hunk 2]
```diff
@@ -70,115 +70,115 @@ class AbstractNormalizations(ABC):
...
     @abstractmethod
-    def time(self) -> u.s:
+    def time(self) -> u.Quantity[u.s]:
         """The normalization for time."""
         ...
 
     @property
     @abstractmethod
-    def wavenumber(self) -> u.m**-1:
+    def wavenumber(self) -> u.Quantity[u.m**1]:
```
Comment (namurphy) [inline plasmapy/simulation/abstractions.py:163]:
```suggestion
    def wavenumber(self) -> u.Quantity[u.m**-1]:
```

File: plasmapy/utils/decorators/checks.py [hunk 3]
```diff
@@ -1184,7 +1184,7 @@ def foo(arg1):
         import astropy.units as u
         from plasmapy.utils.decorators import check_units
 
-        @check_units(arg1={'units': u.K,
+        @check_units(arg1={'units': u.Quantity[u.K],
```
Comment (namurphy) [inline plasmapy/utils/decorators/checks.py:1187]:
```suggestion
        @check_units(arg1={'units': u.K,
```

File: plasmapy/utils/decorators/checks.py [hunk 4]
```diff
@@ -451,7 +451,7 @@ def foo(arg1):
         import astropy.units as u
         from plasmapy.utils.decorators import CheckUnits
 
-        @CheckUnits(arg1={'units': u.K,
+        @CheckUnits(arg1={'units': u.Quantity[u.K],
```
Comment (namurphy) [inline plasmapy/utils/decorators/checks.py:454]:
```suggestion
        @CheckUnits(arg1={'units': u.K,
```

File: plasmapy/utils/decorators/converter.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def angular_freq_to_hz(fn):
 
         @validate_quantities(validations_on_return={'units': [u.rad / u.s, u.Hz]})
         @angular_freq_to_hz
-        def foo(x: u.rad / u.s) -> u.rad / u.s
+        def foo(x: u.Quantity[u.rad / u.s]) -> u.Quantity[u.rad / u.s]
```
Comment (namurphy) [inline plasmapy/utils/decorators/converter.py:50]:
```suggestion
        def foo(x: u.rad / u.s) -> u.rad / u.s
```

File: plasmapy/utils/decorators/tests/test_validators.py [hunk 6]
```diff
@@ -460,7 +460,7 @@ def __init__(self, y: u.cm):
                 self.y = y
 
             @ValidateQuantities(validations_on_return={"can_be_negative": False})
-            def bar(self, x: u.cm) -> u.m:
+            def bar(self, x: u.cm) -> u.Quantity[u.m]:
```
Comment (namurphy) [inline plasmapy/utils/decorators/tests/test_validators.py:463]:
```suggestion
            def bar(self, x: u.cm) -> u.m:
```

File: plasmapy/utils/decorators/validators.py [hunk 7]
```diff
@@ -136,7 +136,7 @@ def foo(arg1):
         import astropy.units as u
         from plasmapy.utils.decorators import ValidateQuantities
 
-        @ValidateQuantities(arg1={'units': u.K,
+        @ValidateQuantities(arg1={'units': u.Quantity[u.K],
```
Comment (namurphy) [inline plasmapy/utils/decorators/validators.py:139]:
```suggestion
        @ValidateQuantities(arg1={'units': u.K,
```

File: plasmapy/utils/decorators/validators.py [hunk 8]
```diff
@@ -527,7 +527,7 @@ def foo(arg1):
         import astropy.units as u
         from plasmapy.utils.decorators import validate_quantities
 
-        @validate_quantities(arg1={'units': u.K,
+        @validate_quantities(arg1={'units': u.Quantity[u.K],
```
Comment (namurphy) [inline plasmapy/utils/decorators/validators.py:530]:
```suggestion
        @validate_quantities(arg1={'units': u.K,
```

File: pyproject.toml [hunk 9]
```diff
@@ -269,7 +269,6 @@ extend-select = [
   "RUF013", # implicit-optional
   "RUF015", # unnecessary-iterable-allocation-for-first-element
   "RUF016", # invalid-index-type
-  "RUF018", # assignment-in-assert
```
Comment (pheuer) [inline pyproject.toml:272]: Why remove this rule?

**Review by pheuer — APPROVED** [review]
Lots of changes, but looks pretty straightforward. I had to open a console to realize that `u.Quantity[u.m]` returns a typing annotation. The coding guide could I guess mention that, but the example is pretty clear. 

Does this PR make this change on every instance in the package?

File: pyproject.toml (hunk 9 above)
Comment (namurphy) [reply pyproject.toml:272]: This is an experimental rule recently added by ruff, and I was locally getting warning messages about not running ruff with ``--preview`` enabled. It's unrelated to this PR.

**Comment by namurphy** [comment]
> Lots of changes, but looks pretty straightforward. I had to open a console to realize that `u.Quantity[u.m]` returns a typing annotation. The coding guide could I guess mention that, but the example is pretty clear.

> Does this PR make this change on every instance in the package?

I used `grep` pretty heavily and I changed all instances of this I could find, but there might be some that are missing.  Enabling mypy in #2424 will help us catch any remaining ones.

Thank you so much for the review!

**Comment by namurphy** [comment]
Correction: this PR doesn't change all of the unit annotations. In `Alfven_speed` and `mass_density`, there are parameters named `density` that accept `Quantity` objects with physical types of number density or mass density. I'm not sure yet how best to do those and make them work with `@validate_quantities`, so I'll leave them for another PR.

**Pull request merged**